### ISSUE TITLE
[Bug]: Source maps now outputs helpers names additionally to input source names

### ISSUE THREAD

**Issue by Gabri3l** [babel/babel#15601]
### 💻

- [ ] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```javascript
async function myfunc() {
  return await someAsyncCall();
}
``` 

### Configuration

```json
{
    "presets": ["env"],
    "sourceMap": true
}
```

### Current and expected behavior

I noticed that with v7.21.0 the result of sourcempas changed. Specifically the names tracked. Given this:

```javascript
const result = babel.transform(src, opts);
console.log(result.map.names);
```

with the given input code I get 155 entries in `result.map.names`. A few of them being: `_regeneratorRuntime`, `Op`, `Object`, `prototype`, `hasOwn`, ...

Before this update I would only get the relevant names from the input code, in this case specifically: `myfunc`, `someAsyncCall`.

### Environment

  System:
    OS: macOS 13.3.1
  Binaries:
    Node: 16.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 8.19.2 - /usr/local/bin/npm
  npmPackages:
    @babel/core: ^7.21.5 => 7.21.5 
    @babel/preset-env: ^7.21.5 => 7.21.5 
    @babel/standalone: ^7.21.7 => 7.21.7

### Possible solution

I might be missing more context but looking at the changelog I didn't notice anything that would highlight this change. I don't necessarily have a good recommendation but if this change was intentional it could be an opt-in via config rather than opt-out.

**EDIT:** On a second look maybe this https://github.com/babel/babel/pull/15022 change?

**Comment by Gabri3l** [comment]
@liuxingbaoyu just curious do you think your change in https://github.com/babel/babel/pull/15022 might have caused this?

**Comment by liuxingbaoyu** [comment]
Yes, this should be the change introduced in this PR. But I'm not sure if this is a BUG, is this causing any problems?

**Comment by Gabri3l** [comment]
With this, it inflates the bundle size up to 70% especially if you have multiple plugins. Not sure this was intentional, I had to revert back to a previous version because the current bundle size is not sustainable for our use case.

**Comment by JLHwung** [comment]
@Gabri3l The source map file `*.map` won't be downloaded to the browser unless users open the developer tools. They are meant to help developers, for example they can map call stacks from production code to source code. I am not sure how larger `*.map` file will impact end users.

**Comment by Gabri3l** [comment]
Yes you're definitely right! For our use case we need the source map as well and I just wanted to make sure this change was intentional since it's a significant size change. I'm not sure it's helpful to have plugins variables as part of the source map, any specific reason why they were added?

I currently have to hack around it and I didn't see this in any change log (I might have missed it though), when I was looking for the reason for this.

**Comment by overlookmotel** [comment]
I have run into the same problem via a different route. Identifiers added to AST programmatically with no source location can end up in `names`.

This does appear to me to be a bug, for 2 reasons:

1. A source map is meant to map to the *source*, so it doesn't make sense to reference names which are not present in the source.
2. It's inconsistent. Sometimes variables appear in `names`, sometimes they don't.

A simple demonstration of the inconsistency:

```js
const {parse} = require('@babel/parser'),
  generate = require('@babel/generator').default;

const ast = parse('function foo() {}', {sourceFilename: '/foo.js'});

const programStatements = ast.program.body;
ast.program.body.unshift({
  type: 'VariableDeclaration',
  kind: 'let',
  declarations: [{
    type: 'VariableDeclarator',
    id: { type: 'Identifier', name: '_x' },
    init: null
  }]
});

const {code: code1, map: map1} = generate(ast, {sourceMaps: true});

console.log(code1);
// -> let _x; function foo() {}
console.log(map1.names);
// -> ['foo']
// `_x` is NOT in `names`

const fnStatements = programStatements[1].body.body;
fnStatements.push({
  type: 'ExpressionStatement',
  expression: { type: 'Identifier', name: '_x' }
});

console.log(code2);
// -> let _x; function foo() { _x; }
console.log(map2.names);
// -> ['foo', '_x']
// Now `_x` IS in `names`
```

This was introduced between 7.20.14 and 7.21.0, so #15022 does seem to be the most obvious cause.

**Comment by liuxingbaoyu** [comment]
Sorry forgot this.

> I'm not sure it's helpful to have plugins variables as part of the source map, any specific reason why they were added?

To be honest, there's no particularly strong need to add them. One possible reason is https://github.com/babel/babel/issues/14907, which can make debugging easier.
For example, a plugin adds the function `xxxhelper` and then the output goes through a minifier. In this case, they can allow devtools to display the original helper name.

Additionally these information may be helpful to `js-debug`. (vs code)

> A source map is meant to map to the source, so it doesn't make sense to reference names which are not present in the source.

Well, I guess you might be right. But unfortunately source maps are crude and difficult to standardize, and many times we have to guess and do things outside of the standard.
https://github.com/babel/babel/pull/15370#issuecomment-1406049596
Here is an even more striking example. We completely agree that this is not a good solution, but it may be the only way we can do it. 🤦‍♂️

> It's inconsistent. Sometimes variables appear in names, sometimes they don't.

This may be a bug, thank you for finding it!

**Comment by overlookmotel** [comment]
OK cool. I see it's complicated!

One small thing: It might be helpful for people to find this issue to correct the typo in the issue title - "Sourcemaps" not "Sourcempas".

**Comment by liuxingbaoyu** [comment]
Sorry I didn't think of a good way.
Can you provide an example?
For example
[sourcemap](https://evanw.github.io/source-map-visualization/#Mjk1AHZhciBUZXN0ID0gLyojX19QVVJFX18qL2Z1bmN0aW9uICgpIHsKICAidXNlIHN0cmljdCI7CgogIGZ1bmN0aW9uIFRlc3QoKSB7CiAgICBiYWJlbEhlbHBlcnMuY2xhc3NDYWxsQ2hlY2sodGhpcywgVGVzdCk7CiAgfQogIGJhYmVsSGVscGVycy5jcmVhdGVDbGFzcyhUZXN0LCBbewogICAga2V5OiAiYmFyIiwKICAgIGdldDogZnVuY3Rpb24gKCkgewogICAgICB0aHJvdyBuZXcgRXJyb3IoIndvdyIpOwogICAgfQogIH1dKTsKICByZXR1cm4gVGVzdDsKfSgpOwp2YXIgdGVzdCA9IG5ldyBUZXN0KCk7CnRlc3QuYmFyOwo2NTEAewogICJ2ZXJzaW9uIjogMywKICAibmFtZXMiOiBbCiAgICAiVGVzdCIsCiAgICAiYmFiZWxIZWxwZXJzIiwKICAgICJjbGFzc0NhbGxDaGVjayIsCiAgICAiY3JlYXRlQ2xhc3MiLAogICAgImtleSIsCiAgICAiZ2V0IiwKICAgICJFcnJvciIsCiAgICAidGVzdCIsCiAgICAiYmFyIgogIF0sCiAgInNvdXJjZXMiOiBbCiAgICAic291cmNlLW1hcHMvY2xhc3MvaW5wdXQuanMiCiAgXSwKICAic291cmNlc0NvbnRlbnQiOiBbCiAgICAiY2xhc3MgVGVzdCB7XG4gIGdldCBiYXIoKSB7XG4gICAgdGhyb3cgbmV3IEVycm9yKFwid293XCIpO1xuICB9XG59XG5cbnZhciB0ZXN0ID0gbmV3IFRlc3Q7XG50ZXN0LmJhcjsiCiAgXSwKICAibWFwcGluZ3MiOiAiSUFBTUEsSUFBSTtFQUFBOztFQUFBLFNBQUFBLEtBQUE7SUFBQUMsWUFBQSxDQUFBQyxjQUFBLE9BQUFGLElBQUE7RUFBQTtFQUFBQyxZQUFBLENBQUFFLFdBQUEsQ0FBQUgsSUFBQTtJQUFBSSxHQUFBO0lBQUFDLEdBQUEsRUFDUixTQUFBQSxDQUFBLEVBQVU7TUFDUixNQUFNLElBQUlDLEtBQUssQ0FBQyxLQUFLLENBQUM7SUFDeEI7RUFBQztFQUFBLE9BQUFOLElBQUE7QUFBQTtBQUdILElBQUlPLElBQUksR0FBRyxJQUFJUCxJQUFJLENBQUQsQ0FBQztBQUNuQk8sSUFBSSxDQUFDQyxHQUFHIgp9)

**Comment by liuxingbaoyu** [comment]
Here's a possible example.
We can notice that the helper name points to a segment that is not an identifier.
Maybe you could try taking advantage of this and check if a segment with an identifier name points to a valid identifier.
I admit this may not be perfect, but that's all I can think of.🤦‍♂️

![image](https://github.com/babel/babel/assets/30521560/de0401c6-0144-4d02-8258-dc1d712510d8)

[sourcemap](https://evanw.github.io/source-map-visualization/#NDM4ACJ1c2Ugc3RyaWN0IjsKCnZhciBfb25lID0gYmFiZWxIZWxwZXJzLmludGVyb3BSZXF1aXJlRGVmYXVsdChyZXF1aXJlKCJvbmUiKSk7CnZhciBfdHdvID0gcmVxdWlyZSgidHdvIik7CnZhciBfdGhyZWUgPSByZXF1aXJlKCJ0aHJlZSIpOwp2YXIgYU5hbWVzcGFjZSA9IGJhYmVsSGVscGVycy5pbnRlcm9wUmVxdWlyZVdpbGRjYXJkKHJlcXVpcmUoImZvdXIiKSk7CmNvbnNvbGUubG9nKF9vbmUuZGVmYXVsdCk7CmNvbnNvbGUubG9nKF90d28uYU5hbWVkKTsKY29uc29sZS5sb2coX3RocmVlLm9yaWcpOwpjb25zb2xlLmxvZyhhTmFtZXNwYWNlKTsKY29uc29sZS5sb2coKDAsIF9vbmUuZGVmYXVsdCkoKSk7CmNvbnNvbGUubG9nKCgwLCBfdHdvLmFOYW1lZCkoKSk7CmNvbnNvbGUubG9nKCgwLCBfdGhyZWUub3JpZykoKSk7CmNvbnNvbGUubG9nKGFOYW1lc3BhY2UoKSk7CjEyNDYAewogICJ2ZXJzaW9uIjogMywKICAibmFtZXMiOiBbCiAgICAiX29uZSIsCiAgICAiYmFiZWxIZWxwZXJzIiwKICAgICJpbnRlcm9wUmVxdWlyZURlZmF1bHQiLAogICAgInJlcXVpcmUiLAogICAgIl90d28iLAogICAgIl90aHJlZSIsCiAgICAiYU5hbWVzcGFjZSIsCiAgICAiaW50ZXJvcFJlcXVpcmVXaWxkY2FyZCIsCiAgICAiY29uc29sZSIsCiAgICAibG9nIiwKICAgICJhRGVmYXVsdCIsCiAgICAiYU5hbWVkIiwKICAgICJhbkFsaWFzZWQiCiAgXSwKICAic291cmNlcyI6IFsKICAgICJtaXNjL3JlZmVyZW5jZS1zb3VyY2UtbWFwL2lucHV0Lm1qcyIKICBdLAogICJzb3VyY2VzQ29udGVudCI6IFsKICAgICJpbXBvcnQgYURlZmF1bHQgZnJvbSBcIm9uZVwiO1xuaW1wb3J0IHsgYU5hbWVkIH0gZnJvbSBcInR3b1wiO1xuaW1wb3J0IHsgb3JpZyBhcyBhbkFsaWFzZWQgfSBmcm9tIFwidGhyZWVcIjtcbmltcG9ydCAqIGFzIGFOYW1lc3BhY2UgZnJvbSBcImZvdXJcIjtcblxuY29uc29sZS5sb2coYURlZmF1bHQpO1xuY29uc29sZS5sb2coYU5hbWVkKTtcbmNvbnNvbGUubG9nKGFuQWxpYXNlZCk7XG5jb25zb2xlLmxvZyhhTmFtZXNwYWNlKTtcblxuY29uc29sZS5sb2coYURlZmF1bHQoKSk7XG5jb25zb2xlLmxvZyhhTmFtZWQoKSk7XG5jb25zb2xlLmxvZyhhbkFsaWFzZWQoKSk7XG5jb25zb2xlLmxvZyhhTmFtZXNwYWNlKCkpOyIKICBdLAogICJtYXBwaW5ncyI6ICI7O0FBQUEsSUFBQUEsSUFBQSxHQUFBQyxZQUFBLENBQUFDLHFCQUFBLENBQUFDLE9BQUE7QUFDQSxJQUFBQyxJQUFBLEdBQUFELE9BQUE7QUFDQSxJQUFBRSxNQUFBLEdBQUFGLE9BQUE7QUFDQSxJQUFBRyxVQUFBLEdBQUFMLFlBQUEsQ0FBQU0sc0JBQUEsQ0FBQUosT0FBQTtBQUVBSyxPQUFPLENBQUNDLEdBQUcsQ0FBQ0MsWUFBUSxDQUFDO0FBQ3JCRixPQUFPLENBQUNDLEdBQUcsQ0FBQ0UsV0FBTSxDQUFDO0FBQ25CSCxPQUFPLENBQUNDLEdBQUcsQ0FBQ0csV0FBUyxDQUFDO0FBQ3RCSixPQUFPLENBQUNDLEdBQUcsQ0FBQ0gsVUFBVSxDQUFDO0FBRXZCRSxPQUFPLENBQUNDLEdBQUcsQ0FBQyxJQUFBQyxZQUFRLEVBQUMsQ0FBQyxDQUFDO0FBQ3ZCRixPQUFPLENBQUNDLEdBQUcsQ0FBQyxJQUFBRSxXQUFNLEVBQUMsQ0FBQyxDQUFDO0FBQ3JCSCxPQUFPLENBQUNDLEdBQUcsQ0FBQyxJQUFBRyxXQUFTLEVBQUMsQ0FBQyxDQUFDO0FBQ3hCSixPQUFPLENBQUNDLEdBQUcsQ0FBQ0gsVUFBVSxDQUFDLENBQUMsQ0FBQyIKfQ==)
